src/sqlfluff/core/rules/std/L010.py

```diff
@@ -45,99 +47,134 @@ class Rule_L010(BaseCrawler):
-                    else:
-                        # If we haven't seen anything yet, then let's default
-                        # to upper
-                        return make_replacement(seg, "upper")
-                # Make a new class and return it.
-                return seg.__class__(raw=new_raw, pos_marker=seg.pos_marker)
-
-            if not seen_case:
-                # Skip this if we haven't seen anything good.
-                # No need to update memory
-                return LintResult(memory=memory)
-            elif (
-                # Are we required to be consistent? (and this is inconsistent?)
-                (
-                    self.capitalisation_policy == "consistent"
-                    and (
-                        # Either because we've seen multiple
-                        (cases_seen and seen_case not in cases_seen)
-                        # Or just because this one is inconsistent internally
-                        or seen_case == "inconsistent"
-                    )
+        # Skip if not an element of the specified type/name
+        if (("type", segment.type) not in self._target_elems) and (
+            ("name", segment.name) not in self._target_elems
+        ):
+            return LintResult(memory=memory)
+
+        # Get the capitalisation policy configuration
+        cap_policy_name = next(
+            k for k in self.config_keywords if k.endswith("capitalisation_policy")
+        )
+        cap_policy = getattr(self, cap_policy_name)
+        cap_policy_opts = [
+            opt
+            for opt in get_config_info()[cap_policy_name]["validation"]
+            if opt != "consistent"
+        ]
+        self.logger.debug(
+            f"Selected '{cap_policy_name}': '{cap_policy}' from options "
+            f"{cap_policy_opts}"
+        )
+
+        refuted_cases = memory.get("refuted_cases", set())
+
+        # Which cases are definitely inconsistent with the segment?
+        if segment.raw[0] != segment.raw[0].upper():
+            refuted_cases.update(["upper", "capitalise", "pascal"])
+            if segment.raw != segment.raw.lower():
+                refuted_cases.update(["lower"])
+        else:
+            refuted_cases.update(["lower"])
+            if segment.raw != segment.raw.upper():
+                refuted_cases.update(["upper"])
+            if segment.raw != segment.raw.capitalize():
+                refuted_cases.update(["capitalise"])
+            if not segment.raw.isalnum():
+                refuted_cases.update(["pascal"])
+
+        # Update the memory
+        memory["refuted_cases"] = refuted_cases
+
+        self.logger.debug(
+            f"Refuted cases after segment '{segment.raw}': {refuted_cases}"
+        )
+
+        # Skip if no inconsistencies, otherwise compute a concrete policy
+        # to convert to.
+        if cap_policy == "consistent":
+            possible_cases = [c for c in cap_policy_opts if c not in refuted_cases]
```

Comment:
nit: `set(cap_policy_opts).difference(refuted_cases)` up to you

Reply:
Will use that!

Reply:
Actually I instead added an additional test that wouldn't pass with your approach.

```
test_consistency_9:
  # Test that UPPER case is chosen first in case of ambiguity
  fail_str: SELECT U, what_Is_This_Case
  fix_str: SELECT U, WHAT_IS_THIS_CASE
```

Basically in case of ambiguity preference should be in the order specified in [`config_info.py`](https://github.com/sqlfluff/sqlfluff/blob/master/src/sqlfluff/core/rules/config_info.py#L53): `["consistent", "upper", "lower", "pascal", "capitalise"]`. In cases of ambiguity, `upper` should take precedence over `pascal` and `capitalise`. 

However, set differences are ordered alphabetically, so this ordering would be lost.